reframe/utility/typecheck.py

```diff
@@ -93,24 +93,35 @@
 import re
 
 
-class _TypeFactory(abc.ABCMeta):
-    def register_subtypes(cls):
-        for t in cls._subtypes:
-            cls.register(t)
+class Type(abc.ABCMeta):
+    def __call__(cls, *args, **kwargs):
+        if (len(args) == 1 and
+            not kwargs and isinstance(args[0], str) and
+            hasattr(cls, '__rfm_cast__')):
+            return cls.__rfm_cast__(*args)
```

Comment:
I think the argument interpretation should be left to the `__rfm_cast__` method.
```suggestion
        if hasattr(cls, '__rfm_cast__'):
            return cls.__rfm_cast__(*args, **kwargs)
```

Reply:
The way you propose it here, `__rfm_cast__` will always supersede the `__init__` constructor, if present. Is this really what we want? This is why I have scoped it here. Also in the way it is implemented now as a `classmethod`, you can't simply create an instance as `C(...)`, because that would lead to an infinite recursion. You'd have to do call explicitly `C.__new__` and `C.__init__`.

Reply:
While I see the value in `__rfm_cast__` interpreting the arguments at will, I don't like that `__init__` has no meaning. I see this functioning as a dynamic dispatch mechanism. What about allowing users to define methods per type they want to convert from and have the dispatching happening here? For example, `__rfm_cast_<type>__` will have the signature `__rfm_cast_type__(x: type) -> type`. Honestly, I don't see a scenario where your conversion function would need more than one argument; and we call it "cast"...

Reply:
I like this `__rfm_cast_<type>__` option. It's far cleaner than potentially having a fat "general" `__rfm_cast__`.

>  I don't see a scenario where your conversion function would need more than one argument; and we call it "cast" anyway...

Yeah, I also think that having this check here makes sense.